ML-289 Add MCMC topic estimation

johnholt opened the pull request:
Restructure for adding MCMC

Add MCMC topic estimation.  MCMC uses Gibbs sampling.  LDA continues to support Expectation Maximization via variational inferencing.  

Signed-off-by: johnholt <john.d.holt@lexisnexis.com>

johnholt commented:
Needs review by @ghalliday @RogerDev @tlhumphrey2   This version adds a Monte Carlo Markov Chain (MCMC) method (aka Gibbs sampling) to estimate the topics.  There are several C++ attributes that create and manage a memory object that is used to track counts in a term-topic table.  The number of terms can be very large, and most terms occur infrequently, so a compressed representation is used for the lower frequency terms.

johnholt commented:
@ghalliday  There is a significant number of C++ attributes that need review.  There are a few normal ones take do some operations on a set of polynomial terms; but more significantly, there are a few that work with a large in memory structure that holds topic by word counts.  Since most words occur fewer times than the number of topics, a sparse matrix representation is used for terms below some frequency threshold.

- diff_counts: find the differences in the topic by term matrix
- init_doc_nd: initialize a document term array
- init_nw: initial term count array
- init_nw_sum: initial term sum array
- mrg_counts: apply the difference counts from the topic by term matrix to a base
- n_phi_log_phi_sum: sum the terms of a polynomial
- resample topics: compute the effect of a document on the topic-term matrix

ghalliday commented on ML/LDA/diff_counts.ecl:
```diff
@@ -0,0 +1,170 @@
+    if (rslt->pairs > diff_pairs) rtlFail(0,"Used pairs exceeds calc");
+    rslt_nwmap[word].list_pos = rslt->pairs;
+    for (size32_t t=0; t<num_topics; t++) {
+      if (work_nw[t] == 0) continue;
```
(when the frequency of the rare terms in both lists is identical)

ghalliday commented on ML/LDA/mrg_counts.ecl:
```diff
@@ -0,0 +1,171 @@
+      if (incr_topic < base_topic) diff_pairs++;
+      if (incr_topic <= base_topic) incr_pos++;
+    }
+    if (incr_pos + 1 < incr_nwmap[word].num_topics) {
```
also not sure why this is +1

ghalliday commented on ML/LDA/resample_topics.ecl:
```diff
@@ -0,0 +1,323 @@
+    //double draw = udis(gen);    // uniform distribution [0,1]
+    double draw = std::rand() * denom * p[num_topics-1];  // scale draw
+    uint32_t topic_ndx;
+//    uint32_t min_pos = 1;
```
Should indicate this is debugging code that has been commented out.

johnholt commented on ML/LDA/diff_counts.ecl:
```diff
@@ -0,0 +1,170 @@
+      if (incr_topic < base_topic) diff_pairs++;
+      if (incr_topic <= base_topic) incr_pos++;
+    }
+    if (incr_pos + 1 < incr_nwmap[word].num_topics) {
```
If we were (lines 79-88) already on the last of the increment data topic for word w, the number of pairs that were different would include the count of the final topic.  If there are topics that we have yet to see (pos+1), those topics add new pairs that must be counted.

Since it took me about 15 minutes to work this out, a comment explaing the counting scheme seems necessary.  I'll add one.

ghalliday commented:
meant to say @johnholt

johnholt commented:
@ghalliday  Added comment blocks to explain the counting object (a combination of dense and sparse arrays) and more comments on how it is used.  I removed all of the typedefs and cleaned up the pointer arithmetic (no void* pointers) to eliminate the warning escalations for the 7.0 release.  The sparse array handling was incorrect, so that has been redone and is not longer fenced off.

ghalliday commented on ML/LDA/diff_counts.ecl:
```diff
@@ -0,0 +1,189 @@
+    // now we have the changes, 0 is no change
+    rslt_nwmap[word].num_topics = 0;
+    rslt_nwmap[word].list_pos = rslt->pairs;
+    for (size32_t t=0; t<num_topics; t++) {
```
minor: Using a local variable for rsltPairs, and then assigning at the end (or assigning at the start and checking at the end) is likely to be more efficient - although the compiler may be able to spot it cannot be aliased.

ghalliday commented on ML/LDA/diff_counts.ecl:
```diff
@@ -0,0 +1,189 @@
+  // cleanup
+  rtlFree(work_nw);
+  if (diff_pairs!=rslt->pairs) {
+    sprintf(buffer, "Error in pairs calculation, %d != %d",
```
pedantic: %u should be used for unsigned.

ghalliday commented on ML/LDA/extract_resample_counts.ecl:
```diff
@@ -0,0 +1,150 @@
+//Extrqact the counts from the memory object after re-sampling the
```
typo extract

ghalliday commented on ML/LDA/mrg_counts.ecl:
```diff
@@ -0,0 +1,180 @@
+  const LDA_MC_Topic_Freq* in_r_nwfreq = (LDA_MC_Topic_Freq*) r_nwfreq;
+  // Step 1, determine the space required.  Space for sparse pairs
+  //is variable.
+  size32_t diff_pairs = 0;
```
picky.  diff_pairs isn't such a good name in this function - it made sense in the function the code was copied from.  merge_pairs probably better.

johnholt replied:
Good idea.  I'll change to mrgd_pairs

ghalliday commented on ML/LDA/mrg_counts.ecl:
```diff
@@ -0,0 +1,180 @@
+                                      (r_ptr + nwmap_offset);
+  LDA_MC_Topic_Freq* rslt_nwfreq = (LDA_MC_Topic_Freq*)
+                                   (r_ptr + nwfreq_offset);
+  // ue a dense array to get counts for each word in turn
```
typo: use

ghalliday commented on ML/LDA/resample_topics.ecl:
```diff
@@ -0,0 +1,304 @@
+      so->nw[i] = in_nw[i].freq;
+    }
+    for (size32_t word = 0; word<rare_words; word++) {
+      uint32_t pos = in_nwmap[word].list_pos;
```
I don't completely follow the code within this for loop.

ghalliday commented on ML/LDA/resample_topics.ecl:
```diff
@@ -0,0 +1,304 @@
+  }
+  // see if we last updated sparse, and if so copy results
+  if (in_z[doc_words-1].ordinal > so->split) {
+    uint32_t last_ordinal = in_z[doc_words-1].ordinal;
```
trivial:  Move up a line, and use in the if test - makes it clearer the array index later on is valid.

johnholt replied:
OK.  I generally like to use the array expression so that there is no doubt, but as you have noted elsewhere this is likely to be more expensive because it is very difficult for the compiler to be certain that the memory location wasn't aliased.

ghalliday commented on ML/LDA/resample_topics.ecl:
```diff
@@ -0,0 +1,304 @@
+      so->nw_c[word] = in_nwmap[word].num_topics;
+    }
+    // Log the allocation
+    sprintf(buffer, "****** Allocated counting object, RID %lld, hand %p",
```
Shouldn't be present in the final version

ghalliday replied:
From embedded c++ you could call

LOGGING_API void LOGGING_CALL logDbgLog(unsigned srcLen, const char * src)

which is defined in the logging plugin.  You will need to make sure that the plugin is linked somehow.

johnholt replied:
Thanks.  I went ahead and wrapped the logging with #ifdef/#endif to drop them but still allow logging to be easily turned on by changing the source.

ghalliday commented on ML/LDA/extract_resample_counts.ecl:
```diff
@@ -0,0 +1,150 @@
+  }
+  // free objects before exit?
+  if (free==true) {
+    char buffer[200];
```
tracing to remove/make conditional in release version?

ghalliday commented:
@johnholt

johnholt commented on ML/LDA/resample_topics.ecl:
```diff
@@ -0,0 +1,304 @@
+        pos++;
+      }
+      // allocate space for each occurrence to have a topic assigned
+      freq = std::min(freq, (int64_t)num_topics); //space for each topic
```
I'll comment to make clearer.  If a word occurs only 3 times and the number of topics is 25, the word can only be in 3 topics.  On the other hand, if the word occurs 30 times and the number of topics is 25, then there can still be just 25 topics.

johnholt replied:
The in_nwmap num_topics values for the word show the number of topics in the last re-sampling.  This time it might be different.  
In the three occurrence case, those three words may have been associated with only 2 topics in the last run.  This time they could be associated with three topics.  So, we always noeed enough room for the maximum number of topics.
When the number of topics is large (thousands), most of the words do not occur as many times as there are topics, so these words will have fewer topics than the number of topics.

johnholt commented:
@ghalliday  I made the changes suggested in the reviews including wrapping the logging with #ifdef/#endif so that it is not normally enabled.  I also found a few more attributes that needed to be fixed because of the warning escalation.

johnholt commented:
@ghalliday  @richardkchapman  This should be good to merge.

richardkchapman commented:
@ghalliday I will leave that determination to you
